heron/cgi-bin/heron-wsgi.py

```diff
@@ -0,0 +1,397 @@
+        # open('files/' + fn, 'wb').write(fileitem.file.read())
+        # Echo back file content to client
+        # print tempfile.gettempprefix()
+        # print file_item.filename
+        # file_path = os.path.join(self.path, file_item.filename)
+        # temp_file = tempfile.TemporaryFile()
+        # file_path = file_item.file.name
+        data = file_item.value
+
+        # if the upload is a .zip file we assume a zipped ESRI Shapefile
+        # we convert it to GeoJSON, such that the client can read it
+        # The config in the Heron client (Upload or Editor) should then have an entry like:
+        # {name: 'ESRI Shapefile (1 laag, gezipped)', fileExt: '.zip', mimeType: 'text/plain', formatter: 'OpenLayers.Format.GeoJSON'}
+        f, file_ext_in = os.path.splitext(file_item.filename.lower())
+        if file_ext_in == '.zip' or file_ext_in == '.gpkg' or file_ext_in == '.csv' or 'target_srs' in params:
+            # Convert with ogr2ogr
+            work_dir = prepare_dir(suffix='_upwrk')
+
+            in_file, out_file = prepare_upload_files(file_item, file_ext_in, work_dir)
+            assign_srs = params.getvalue('assign_srs', None)
+            source_srs = params.getvalue('source_srs', None)
+            target_srs = params.getvalue('target_srs', None)
+
+            out_file = ogr2ogr(out_file, in_file, 'GeoJSON', assign_srs=assign_srs, source_srs=source_srs, target_srs=target_srs)
+            data = get_file_data(out_file)
+            shutil.rmtree(work_dir)
+
+        if encoding == 'base64':
+            data = base64.b64encode(data)
+        elif encoding == 'url':
+            data = urllib.quote(data)
+        elif encoding == 'escape':
+            data = cgi.escape(data)
+    else:
+        data = 'No file data received'
+
+    status = '200 OK'
+    response_headers = [('Content-type', mime),("Content-Length", len(data))]
+    return (status, response_headers, data)
+
+# Action handlers: jump table with function pointers
+HANDLERS = {
+    'download': download,
+    'upload': upload
+}
+
+def application(environ, start_response):
+    global params
+
+
+    if "REQUEST_METHOD" in environ and environ['REQUEST_METHOD'] == 'GET':
+        get_env = environ.copy()
+        params = cgi.FieldStorage(
+            qs=environ["QUERY_STRING"],
+            environ=get_env,
+            keep_blank_values=True
+        )
+    
+    if "REQUEST_METHOD" in environ and environ['REQUEST_METHOD'] == 'POST':
+        post_env = environ.copy()
+        post_env['QUERY_STRING'] = ''
+        params = cgi.FieldStorage(
+            fp=environ['wsgi.input'],
+            environ=post_env,
+            keep_blank_values=True
+        )
+
+    # Execute function based on 'action' param
+    status, response_headers, data = HANDLERS[params.getvalue('action', 'download')]()
+    
+    # return response 
+    start_response(status, response_headers)
+    return [data]
+
+if __name__ == '__main__':
```

Comment:
Just,

Sometimes wsgi is better, sometimes cgi. That's what I like about open 
source. The freedom of choice.

Indeed, I agree on having a common code base for both cgi and wsgi. That 
actually shouldn't be to difficult, but we do run the risk that if we 
make a heron.cgi and a heron.wsgi which import a heron-base, the setup 
will become more difficult for the novice user.

Maybe we can solve this by creating one code base and then build a cgi 
and a wsgi version from it. Simple for the user (either use heron.cgi or 
heron.wsgi) and simple to maintain due to the common code base.

I do have -limited- time to work on this together.

MArco




On 15-10-15 13:01, Just van den Broecke wrote:
>
> In heron/cgi-bin/heron-wsgi.py 
> <https://github.com/heron-mc/heron-mc/pull/433#discussion_r42109556>:
>
> > +        post_env = environ.copy()
> > +        post_env['QUERY_STRING'] = ''
> > +        params = cgi.FieldStorage(
> > +            fp=environ['wsgi.input'],
> > +            environ=post_env,
> > +            keep_blank_values=True
> > +        )
> > +
> > +    # Execute function based on 'action' param
> > +    status, response_headers, data = HANDLERS[params.getvalue('action', 'download')]()
> > +
> > +    # return response
> > +    start_response(status, response_headers)
> > +    return [data]
> > +
> > +if __name__ == '__main__':
>
> Thanks for this work Marco! I think that using WSGI is cleaner. But 
> will also require extra installs, in Python e.g. |mod_wsgi| on Apache. 
> |heron.cgi| is already very challenging for users. Raw CGI is 
> sometimes easier. I think it will be optimal to have both a CGI and 
> WSGI version. Only I would like to see a single common codebase as in 
> |heron-wsgi.py| code is duplicated which will make maintenance harder 
> and causing confusion.
>
> The current code in |heron.cgi| is not very clean (was my first Python 
> program) so this is a good opportunity. If possible I would like to 
> work towards a single |heron.cgi| that supports both raw cgi and wsgi. 
> I don't know if that is possible. Other option is to have a single 
> |"heronogr.py"| with all "OGR" logic and two small scripts heron.cgi 
> and heron-wsgi.cgi that call on the logic from heronogr.py.
>
> If you agree also to work further with on the above proposal I shall 
> first merge this PR and from then work together on the 
> redesign/implementation.
>
> BTW in the further future we should be able to call |ogr2ogr| via 
> Python i.s.o. launching the |ogr2ogr| command see 
> http://erouault.blogspot.nl/2015/10/gdal-and-ogr-utilities-as-library.html
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/heron-mc/heron-mc/pull/433/files#r42109556>.
>